### PR TITLE
New bot: detect red-pill contracts

### PR DESCRIPTION
**Pull request by apehex**
Here's the second bot of the "evasion-detection" flavor!
It detects "red-pill" contracts that detect simulation-env by checking the global variables against default values.
The detection happens on contract creation by static analysis.

### CODE REVIEW

File: red-pill-contracts-py/README.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  - `from`: the transaction sender
+  - `to`: the transaction recipient
+  - `anomaly_score`: the alert rate for this combination of bot / alert type
+
+## Options
+
+The bot settings are located in `src/options.py`:
+
+```python
+MIN_CONFIDENCE = 0.7 # probability threshold
```
Comment (christian-forta) [inline red-pill-contracts-py/README.md:51]: I would recommend to elaborate a bit on how you handle probabilities. Given conceptually this detection is a heuristic, this doesnt make sense. Knowing how the code works, I understand. Maybe just add a comment that there will be heuristics applied on the code, which each represent a weight. These are added to calculate a confidence from 0-1 (is the score normalized?).

File: red-pill-contracts-py/README.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ALERT_HISTORY_SIZE = 16384 # in number of transactions recorded
```
Comment (christian-forta) [inline red-pill-contracts-py/README.md:52]: in number of alerts?

File: red-pill-contracts-py/src/agent.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+# CONSTANTS ###################################################################
+
+CHAIN_ID = 1
+PROVIDER = Web3(Web3.HTTPProvider(get_json_rpc_url()))
+
+# INIT ########################################################################
+
+forta_toolkit.logging.setup_logger(logging.INFO)
+forta_toolkit.parsing.env.load_secrets()
```
Comment (christian-forta) [inline red-pill-contracts-py/src/agent.py:33]: pls remove if not required

File: red-pill-contracts-py/src/agent.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    def __handle_transaction(log: TransactionEvent) -> list:
+        """Main function called on the logs gathered by the Forta network."""
+        global CHAIN_ID
+        # result: list of alerts
+        __findings = []
+        # parse all the data
+        __tx = forta_toolkit.parsing.transaction.parse_transaction_data(transaction=log.transaction)
+        __logs = [forta_toolkit.parsing.logs.parse_log_data(log=__l) for __l in log.logs]
+        __traces = [forta_toolkit.parsing.traces.parse_trace_data(trace=__t) for __t in log.traces]
+        # iterate over event logs
```
Comment (christian-forta) [inline red-pill-contracts-py/src/agent.py:66]: remove?

File: red-pill-contracts-py/src/agent.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+# TODO ########################################################################
+
+# import forta_toolkit.parsing.transaction
+
+# import ioseeth.metrics.evasion.morphing.logic_bomb
+# import ioseeth.metrics.evasion.redirection
+
+# is_hidden_proxy = functools.lru_cache(maxsize=128)(ioseeth.metrics.evasion.redirection.is_hidden_proxy)
+# is_red_pill = functools.lru_cache(maxsize=128)(ioseeth.metrics.evasion.morphing.logic_bomb.is_red_pill)
+
```
Comment (christian-forta) [inline red-pill-contracts-py/src/agent.py:106]: remove?

File: red-pill-contracts-py/src/findings.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
+"""Format the agent findings into Forta alerts"""
```
Comment (christian-forta) [inline red-pill-contracts-py/src/findings.py:1]: this file contains a bunch of items unrelated to redpill. can those be removed?

File: red-pill-contracts-py/src/scoring.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+is_trace_factory_contract_creation = functools.lru_cache(maxsize=128)(imemm.is_trace_factory_contract_creation)
+is_trace_mutant_contract_creation = functools.lru_cache(maxsize=128)(imemm.is_trace_mutant_contract_creation)
+is_trace_red_pill_contract_creation = functools.lru_cache(maxsize=128)(imeml.is_trace_red_pill_contract_creation)
+
+def score_trace(trace: dict, **kwargs) -> dict:
+    """Estimate the probabilities that the contract performs evasion techniques."""
+    # scores each evasion technique
+    __scores = {(sf.EvasionTechnique.Metamorphism, sf.MetamorphismAlert.FactoryDeployment): 0.5, (sf.EvasionTechnique.Metamorphism, sf.MetamorphismAlert.MutantDeployment): 0.5, (sf.EvasionTechnique.LogicBomb, sf.LogicBombAlert.RedPill): 0.5}
+    # update scores
```
Comment (christian-forta) [inline red-pill-contracts-py/src/scoring.py:21]: is this where you configure the bot just for the red pill?

File: red-pill-contracts-py/tests/test_agent.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
+"""Test the agent on a fork"""
```
Comment (christian-forta) [inline red-pill-contracts-py/tests/test_agent.py:1]: you mentioned on chat these tests are no good? should we remove those?

**Review by christian-forta — APPROVED** [review]

File: red-pill-contracts-py/README.md (hunk 2 above)
Comment (apehex) [reply red-pill-contracts-py/README.md:97]: Nope transactions :D
The history has an empty record for transactions without alerts in order to calculate the "alert-rate"

File: red-pill-contracts-py/src/scoring.py (hunk 7 above)
Comment (apehex) [reply red-pill-contracts-py/src/scoring.py:17]: it's still extra code, i removed it

File: red-pill-contracts-py/tests/test_agent.py (hunk 8 above)
Comment (apehex) [reply red-pill-contracts-py/tests/test_agent.py:1]: i updated them, i just miss an actual transaction log of a red-pill contract

File: red-pill-contracts-py/src/findings.py (hunk 6 above)
Comment (apehex) [reply red-pill-contracts-py/src/findings.py:1]: yup

File: red-pill-contracts-py/src/agent.py (hunk 5 above)
Comment (apehex) [reply red-pill-contracts-py/src/agent.py]: alright!

File: red-pill-contracts-py/src/agent.py (hunk 4 above)
Comment (apehex) [reply red-pill-contracts-py/src/agent.py]: not all but yep :)

File: red-pill-contracts-py/src/agent.py (hunk 3 above)
Comment (apehex) [reply red-pill-contracts-py/src/agent.py]: less go

File: red-pill-contracts-py/README.md (hunk 1 above)
Comment (apehex) [reply red-pill-contracts-py/README.md:159]: True, there's a gap ! I documented this

**Comment by apehex** [comment]
I just force pushed to get back on top of the main branch ^^

**Pull request merged**